File: Runtime/Scripts/Events/VREvent.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using System.Runtime.Serialization;

namespace IVLab.MinVR3
{
    [System.Serializable]
    public class VREvent : ISerializable
    {
        public VREvent(string eventName) : this()
        {
            m_Name = eventName;
        }

        public VREvent()
        {
            m_DataTypeName = ""; // by default, there is no data in a VREvent
        }

        public string name {
            get => m_Name;
            set => m_Name = value;
        }

        public string GetName()
        {
            return m_Name;
        }

        public string GetDataTypeName()
        {
            return m_DataTypeName;
        }

        public T GetData<T>()
        {
            return (this as VREventT<T>).data;
        }

        public bool Matches(IVREventPrototype eventPrototype)
        {
            return (GetName() == eventPrototype.GetEventName()) &&
                (GetDataTypeName() == eventPrototype.GetEventDataTypeName());
        }

        /// <summary>
        /// Does a deep copy of the VREvent, subclasses override this to preserve the correct type of event.
        /// </summary>
        /// <returns></returns>
        public virtual VREvent Clone()
        {
            return new VREvent(m_Name);
        }


        public static VREvent CreateFromJson(string eventJson)
        {
            try {
                // Serialize once to get base fields of VREvent, including the type
                VREvent evt = JsonUtility.FromJson<VREvent>(eventJson);

                // PLACE 2 OF 2 TO MODIFY WHEN ADDING A NEW DATA TYPE
                // Convert to the actual type
                // There may be a better way to do this, but at least a switch is fast.
                // Alternative would be to use reflection to convert the type
                // using the s_AvailableDataTypes dictionary.
                //
                // Some ugliness for parsing built-in types here due to Unity not
                // supporting .NET 7 yet, which has an `IParsable<>` interface.
                System.Text.RegularExpressions.GroupCollection matchGroups;
                switch (evt.GetDataTypeName()) {
                    case "Vector2":
                        evt = JsonUtility.FromJson<VREventVector2>(eventJson);
                        break;
                    case "Vector3":
                        evt = JsonUtility.FromJson<VREventVector3>(eventJson);
                        break;
                    case "Vector4":
                        evt = JsonUtility.FromJson<VREventVector4>(eventJson);
                        break;
                    case "Quaternion":
                        evt = JsonUtility.FromJson<VREventQuaternion>(eventJson);
                        break;
                    case "GameObject":
                        evt = JsonUtility.FromJson<VREventGameObject>(eventJson);
                        break;
                    case "Single":
                        // try to parse, assuming data value is inside an object under the field named "value"
                        matchGroups = builtinTypeRegexForDataInsideObject.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            evt = new VREventFloat(evt.name, Single.Parse(matchGroups[1].Value));
                            break;
                        }
                        // try to parse, assuming data value is in a simple name/value pair
                        matchGroups = dataRegexForIntOrFloat.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            evt = new VREventFloat(evt.name, Single.Parse(matchGroups[1].Value));
                            break;
                        }

                        Debug.Log("Problem extracting data from serialized VREvent.");
                        break;
                    case "Int32":
                        // try to parse, assuming data value is inside an object under the field named "value"
                        matchGroups = builtinTypeRegexForDataInsideObject.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            evt = new VREventInt(evt.name, Int32.Parse(matchGroups[1].Value));
                            break;
                        }
                        // try to parse, assuming data value is in a simple name/value pair
                        matchGroups = dataRegexForIntOrFloat.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            evt = new VREventInt(evt.name, Int32.Parse(matchGroups[1].Value));
                            break;
                        }
                        Debug.Log("Problem extracting data from serialized VREvent.");
                        break;
                    case "String":
                        // try to parse, assuming data value is inside an object under the field named "value"
                        matchGroups = builtinTypeRegexForDataInsideObject.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            // remove start/end quotes
                            String stringValue = matchGroups[1].Value.Substring(1, matchGroups[1].Value.Length - 2);
                            evt = new VREventString(evt.name, stringValue);
                            break;
                        }
                        // try to parse, assuming data value is in a simple name/value pair
                        matchGroups = dataRegexForQuotedString.Match(eventJson).Groups;
                        if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                            if (matchGroups[1].Value[0] != '{') {
                                // normal quoted string
                                evt = new VREventString(evt.name, stringValue);
                                break;
                            } else {
                                // string starts with { so assume we have a json string and we need a more complex regex to get it all
                                matchGroups = dataRegexForQuotedJson.Match(eventJson).Groups;
                                if ((matchGroups[0].Success) && (matchGroups[1].Value != String.Empty)) {
                                    evt = new VREventString(evt.name, stringValue);
                                    break;
                                }
                            }
                        }
                        Debug.Log("Problem extracting data from serialized VREvent.");
                        break;
                    default:
                        break;
                }

                return evt;
            } catch (System.Exception exc) {
                Debug.LogError("Unable to deserialize JSON VREvent message:\n" + exc);
                return null;
            }
        }

        public override string ToString()
        {
            return $"{GetName()} ()";
        }

        /// <summary>
        /// True for raw VREvents created from input devices, trackers, etc.  These need to be sent
        /// across the network to synchronize nodes when running in cluster mode so that all nodes
        /// receive the same input at the same time.  It is also possible to use VREvents for message
        /// passing within an application, and in that case, the data might reference some local object
        /// like a GameObject.  These are considered secondary (i.e., application-specific) events,
        /// and they should not be sent across the network because the reference to a local object
        /// will not transfer to another cluster node running a separate copy of the application.
        /// Instead, secondary events should be regenerated from the primary events by running the same
        /// secondary-event-producing script on each node in the cluster.
        /// </summary>
        public virtual bool IsClusterSafe()
        {
            return true;
        }

        protected VREvent(SerializationInfo info, StreamingContext context) : this()
        {
            m_Name = info.GetString("name");
            m_DataTypeName = info.GetString("dataTypeName");
        }

        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue("name", m_Name);
            info.AddValue("dataTypeName", m_DataTypeName);
        }

        [SerializeField]
        protected string m_Name;

        [SerializeField]
        protected string m_DataTypeName;

        // PLACE 1 OF 2 TO MODIFY WHEN ADDING A NEW DATA TYPE
        public static Dictionary<string, Type> AvailableDataTypes { get; } = new Dictionary<string, Type>()
        {
            { "", typeof(VREvent) },
            { typeof(Vector2).Name, typeof(VREventVector2) },
            { typeof(Vector3).Name, typeof(VREventVector3) },
            { typeof(Vector4).Name, typeof(VREventVector4) },
            { typeof(Quaternion).Name, typeof(VREventQuaternion) },
            { typeof(float).Name, typeof(VREventFloat) },
            { typeof(int).Name, typeof(VREventInt) },
            { typeof(string).Name, typeof(VREventString) },
            { typeof(GameObject).Name, typeof(GameObject) },
        };

        // dfk comments 7/14/23:
        // this regex seems to expect the data to come in as an object with a field named value, i.e., "m_Data":{"value":"datavalue"}.  Is this what the 
        // websocket json serializer generates?  it does not seem to be what is generated by the JsonUtility used in this class.
        private static System.Text.RegularExpressions.Regex builtinTypeRegexForDataInsideObject = new System.Text.RegularExpressions.Regex(@"""m_Data"":{""value"":([\s\S]+)}}");

        // the JsonUtility used in this class as well as the jsoncpp package used in the forceserver serialize m_Data as a simple name/value pair
        // i.e., "m_Data":datavalue.  The following regex works for datavalues followed by a , or a }, which is good for ints and floats.
        private static System.Text.RegularExpressions.Regex dataRegexForIntOrFloat = new System.Text.RegularExpressions.Regex(@"""m_Data"":(.*?)[,}]");
        
        // extracts the string inside quotes that comes after "m_Data": handling any escaped quotes within the string
        private static System.Text.RegularExpressions.Regex dataRegexForQuotedString = new System.Text.RegularExpressions.Regex(@"""m_Data"":""((?:[^""\\]|\\.)*)""");
        
        // extracts the largest sequence of matched {} inside the quoted string that comes after "m_Data":
        private static System.Text.RegularExpressions.Regex dataRegexForQuotedJson = new System.Text.RegularExpressions.Regex(@"""m_Data"":""(\{(?>[^}{]+|(?<c>)\{|(?<-c>)})+})""");
        
        
        
    }

} // namespace

File: MinVR3.cpp~/src/vr_event.cpp

#include "vr_event.h"

#include <string>
#include <iostream>

#include "json/json.h"


VREvent::VREvent(const std::string& event_name) {
	name_ = event_name;
	data_type_name_ = "";
}

VREvent::VREvent(const std::string& event_name, const std::string& data_type_name) {
	name_ = event_name;
	data_type_name_ = data_type_name;
}

VREvent::VREvent() {
	name_ = "";
	data_type_name_ = "";
}

VREvent::~VREvent() {}

std::string VREvent::get_name() const {
	return name_;
}

std::string VREvent::get_data_type_name() const {
	return data_type_name_;
}

void VREvent::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
}

std::string VREvent::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}

VREvent* VREvent::CreateFromJson(const std::string& eventJsonStr) {
	Json::Reader reader;
	Json::Value eventJson;
	if (!reader.parse(eventJsonStr, eventJson)) {
		std::cerr << reader.getFormattedErrorMessages() << std::endl;
		return NULL;
	}

	std::string name = eventJson["m_Name"].asString();
	std::string data_type_name = eventJson["m_DataTypeName"].asString();

	// empty string means no data payload with the event
	if (data_type_name == "") {
		return new VREvent(name);
	}

	// else, different cases based on the data type:
	Json::Value data = eventJson["m_Data"];
	if (data_type_name == "Vector2") {
		return new VREventVector2(name, data["x"].asFloat(), data["y"].asFloat());
	}
	else if (data_type_name == "Vector3") {
		return new VREventVector3(name, data["x"].asFloat(), data["y"].asFloat(), data["z"].asFloat());
	}
	else if (data_type_name == "Vector4") {
		return new VREventVector4(name, data["x"].asFloat(), data["y"].asFloat(), data["z"].asFloat(), data["w"].asFloat());
	}
	else if (data_type_name == "Quaternion") {
		return new VREventQuaternion(name, data["x"].asFloat(), data["y"].asFloat(), data["z"].asFloat(), data["w"].asFloat());
	}
    else if (data_type_name == "String") {
		return new VREventString(name, data.asString());
	}
	else if (data_type_name == "Int32") {
		return new VREventInt(name, data.asInt());
	}
	else if (data_type_name == "Single") {
		return new VREventFloat(name, data.asFloat());
	}

	std::cerr << "VREvent::FromJson() unknown event data type: " << data_type_name << std::endl;
	return NULL;
}


VREventInt::VREventInt(const std::string& eventName, int data) : VREvent(eventName, "Int32") {
    data_ = data;
}

VREventInt::VREventInt() : VREvent("", "Int32") {
    data_ = 0;
}

VREventInt::~VREventInt() {}

int VREventInt::get_data() const { return data_; }

void VREventInt::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    data_ = eventJson["m_Data"].asInt();
}

std::string VREventInt::ToJson() const {
    Json::Value eventJson;
    eventJson["m_Name"] = name_;
    eventJson["m_DataTypeName"] = data_type_name_;
    eventJson["m_Data"] = data_;
    Json::FastWriter fastWriter;
    std::string eventJsonStr = fastWriter.write(eventJson);
    return eventJsonStr;
}


VREventFloat::VREventFloat(const std::string& eventName, float data) : VREvent(eventName, "Single") {
    data_ = data;
}

VREventFloat::VREventFloat() : VREvent("", "Single") {
    data_ = 0.0f;
}

VREventFloat::~VREventFloat() {}

float VREventFloat::get_data() const { return data_; }

void VREventFloat::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    data_ = eventJson["m_Data"].asFloat();
}

std::string VREventFloat::ToJson() const {
    Json::Value eventJson;
    eventJson["m_Name"] = name_;
    eventJson["m_DataTypeName"] = data_type_name_;
    eventJson["m_Data"] = data_;
    Json::FastWriter fastWriter;
    std::string eventJsonStr = fastWriter.write(eventJson);
    return eventJsonStr;
}


VREventVector2::VREventVector2(const std::string& eventName, float x, float y) : VREvent(eventName, "Vector2") {
	x_ = x;
	y_ = y;
}

VREventVector2::VREventVector2() : VREvent("", "Vector2") {
	x_ = 0;
	y_ = 0;
}

VREventVector2::~VREventVector2() {}

std::vector<float> VREventVector2::get_data() const {
    return std::vector<float>{x_, y_};
}

float VREventVector2::x() const { return x_; }
float VREventVector2::y() const { return y_; }

void VREventVector2::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    x_ = eventJson["m_Data"]["x"].asFloat();
    y_ = eventJson["m_Data"]["y"].asFloat();
}

std::string VREventVector2::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	Json::Value dataJson;
	dataJson["x"] = x_;
	dataJson["y"] = y_;
	eventJson["m_Data"] = dataJson;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}


VREventVector3::VREventVector3(const std::string& eventName, float x, float y, float z) : VREvent(eventName, "Vector3") {
	x_ = x;
	y_ = y;
	z_ = z;
}

VREventVector3::VREventVector3() : VREvent("", "Vector3") {
	x_ = 0;
	y_ = 0;
	z_ = 0;
}

VREventVector3::~VREventVector3() {}

std::vector<float> VREventVector3::get_data() const {
    return std::vector<float>{x_, y_, z_};
}

float VREventVector3::x() const { return x_; }
float VREventVector3::y() const { return y_; }
float VREventVector3::z() const { return z_; }

void VREventVector3::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    x_ = eventJson["m_Data"]["x"].asFloat();
    y_ = eventJson["m_Data"]["y"].asFloat();
    z_ = eventJson["m_Data"]["z"].asFloat();
}

std::string VREventVector3::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	Json::Value dataJson;
	dataJson["x"] = x_;
	dataJson["y"] = y_;
	dataJson["z"] = z_;
	eventJson["m_Data"] = dataJson;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}


VREventVector4::VREventVector4(const std::string& eventName, float x, float y, float z, float w) : VREvent(eventName, "Vector4") {
	x_ = x;
	y_ = y;
	z_ = z;
	w_ = w;
}

VREventVector4::VREventVector4() : VREvent("", "Vector4") {
	x_ = 0;
	y_ = 0;
	z_ = 0;
	w_ = 0;
}

VREventVector4::~VREventVector4() {}

std::vector<float> VREventVector4::get_data() const {
    return std::vector<float>{x_, y_, z_, w_};
}

float VREventVector4::x() const { return x_; }
float VREventVector4::y() const { return y_; }
float VREventVector4::z() const { return z_; }
float VREventVector4::w() const { return w_; }

void VREventVector4::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    x_ = eventJson["m_Data"]["x"].asFloat();
    y_ = eventJson["m_Data"]["y"].asFloat();
    z_ = eventJson["m_Data"]["z"].asFloat();
    w_ = eventJson["m_Data"]["w"].asFloat();
}

std::string VREventVector4::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	Json::Value dataJson;
	dataJson["x"] = x_;
	dataJson["y"] = y_;
	dataJson["z"] = z_;
	dataJson["w"] = w_;
	eventJson["m_Data"] = dataJson;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}


VREventQuaternion::VREventQuaternion(const std::string& eventName, float x, float y, float z, float w) : VREvent(eventName, "Quaternion") {
	x_ = x;
	y_ = y;
	z_ = z;
	w_ = w;
}

VREventQuaternion::VREventQuaternion() : VREvent("", "Quaternion") {
	x_ = 0;
	y_ = 0;
	z_ = 0;
	w_ = 0;
}

VREventQuaternion::~VREventQuaternion() {}

std::vector<float> VREventQuaternion::get_data() const {
    return std::vector<float>{x_, y_, z_, w_};
}

float VREventQuaternion::x() const { return x_; }
float VREventQuaternion::y() const { return y_; }
float VREventQuaternion::z() const { return z_; }
float VREventQuaternion::w() const { return w_; }

void VREventQuaternion::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    x_ = eventJson["m_Data"]["x"].asFloat();
    y_ = eventJson["m_Data"]["y"].asFloat();
    z_ = eventJson["m_Data"]["z"].asFloat();
    w_ = eventJson["m_Data"]["w"].asFloat();
}

std::string VREventQuaternion::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	Json::Value dataJson;
	dataJson["x"] = x_;
	dataJson["y"] = y_;
	dataJson["z"] = z_;
	dataJson["w"] = w_;
	eventJson["m_Data"] = dataJson;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}


VREventString::VREventString(const std::string& eventName, const std::string& data) : VREvent(eventName, "String") {
	str_ = data;
}

VREventString::VREventString() : VREvent("", "String") {
	str_ = "";
}

VREventString::~VREventString() {}

std::string VREventString::get_data() const { return str_; }

void VREventString::SetFromJson(const std::string &eventJsonStr) {
    Json::Reader reader;
    Json::Value eventJson;
    if (!reader.parse(eventJsonStr, eventJson)) {
        std::cerr << reader.getFormattedErrorMessages() << std::endl;
        return;
    }
    name_ = eventJson["m_Name"].asString();
    data_type_name_ = eventJson["m_DataTypeName"].asString();
    str_ = eventJson["m_Data"].asString();
}

std::string VREventString::ToJson() const {
	Json::Value eventJson;
	eventJson["m_Name"] = name_;
	eventJson["m_DataTypeName"] = data_type_name_;
	eventJson["m_Data"] = str_;
	Json::FastWriter fastWriter;
	std::string eventJsonStr = fastWriter.write(eventJson);
	return eventJsonStr;
}


void VREvent::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "]";
}

void VREventInt::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = " << get_data();
}

void VREventFloat::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = " << get_data();
}

void VREventVector2::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = (" << x() << ", " << y() << ")";
}

void VREventVector3::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = (" << x() << ", " << y() << ", " << z() << ")";
}

void VREventVector4::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = (" << x() << ", " << y() << ", " << z() << ", " << w() << ")";
}

void VREventQuaternion::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = (" << x() << ", " << y() << ", " << z() << ", " << w() << ")";
}

void VREventString::Print(std::ostream& os) const {
    os << get_name() << " [" << get_data_type_name() << "] = '" << get_data() + "'";
}


std::ostream & operator<< (std::ostream &os, const VREvent &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventInt &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventFloat &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventVector2 &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventVector3 &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventVector4 &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventQuaternion &e) {
    e.Print(os);
    return os;
}

std::ostream & operator<< ( std::ostream &os, const VREventString &e) {
    e.Print(os);
    return os;
}
